### PR TITLE
Adds opentracing and jaeger support

### PR DESCRIPTION
**Pull request by jmprusi**
This PR adds opentracing and jaeger support to the runtime and build
image.

### CODE REVIEW

File: scripts/build-opentracing.sh [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+mkdir /opt/app-root/src/opentracing-libs/
+cp objs/ngx_http_opentracing_module.so /opt/app-root/src/nginx-modules/
+cp objs/ngx_http_jaeger_module.so /opt/app-root/src/nginx-modules/
+
+# Move libraries
+cp /usr/local/lib/libjaeger* /opt/app-root/src/opentracing-libs/
+cp /usr/local/lib/libopentracing* /opt/app-root/src/opentracing-libs/
+
+# clean
+rm -rf /tmp/*
```
Comment (mikz) [inline scripts/build-opentracing.sh]: This is quite nasty no?

File: scripts/build-opentracing.sh [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+rm -rf /opt/app-root/src/.hunter
```
Comment (mikz) [inline scripts/build-opentracing.sh]: This can be `~/.hunter` ?

File: scripts/build-opentracing.sh [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+# Move modules to another dir.
+mkdir /opt/app-root/src/nginx-modules/
+mkdir /opt/app-root/src/opentracing-libs/
+cp objs/ngx_http_opentracing_module.so /opt/app-root/src/nginx-modules/
+cp objs/ngx_http_jaeger_module.so /opt/app-root/src/nginx-modules/
+
+# Move libraries
+cp /usr/local/lib/libjaeger* /opt/app-root/src/opentracing-libs/
+cp /usr/local/lib/libopentracing* /opt/app-root/src/opentracing-libs/
```
Comment (mikz) [inline scripts/build-opentracing.sh]: Could those be in the `/opt/app-root/lib` ?

Comment (jmprusi) [reply scripts/build-opentracing.sh]: Yes

File: scripts/build-opentracing.sh [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+md5sum -c <<<"${OPENRESTY_MD5} /tmp/openresty.tar.gz"
+tar zxf /tmp/openresty.tar.gz -C /tmp/
+
+git clone -b "$OPENTRACING_CPP_VERSION" https://github.com/opentracing/opentracing-cpp.git /tmp/opentracing
+git clone -b "$NGINX_OPENTRACING_VERSION" https://github.com/opentracing-contrib/nginx-opentracing.git /tmp/nginx-opentracing
+git clone -b "$JAEGER_CPP_VERSION" https://github.com/jaegertracing/cpp-client.git /tmp/jaeger-cpp
+git clone -b "${OPENSSL_VERSION}" https://github.com/openssl/openssl.git /tmp/openssl
+
+###################### fix
+export RESTY_PCRE_VERSION="8.41"
```
Comment (mikz) [inline scripts/build-opentracing.sh]: Would `-devel` package of https://github.com/openresty/openresty-packaging/blob/master/rpm/SPECS/openresty-pcre.spec help?

File: scripts/build-opentracing.sh (hunk 1 above)
Comment (jmprusi) [reply scripts/build-opentracing.sh]: not really elegant.. but what do you propose?

File: scripts/build-opentracing.sh (hunk 2 above)
Comment (jmprusi) [reply scripts/build-opentracing.sh]: yes

File: scripts/build-opentracing.sh (hunk 4 above)
Comment (mikz) [reply scripts/build-opentracing.sh]: I see it is installed, so why do we need this?

Comment (jmprusi) [reply scripts/build-opentracing.sh]: I will try to reuse those one.

File: scripts/build-opentracing.sh (hunk 1 above)
Comment (mikz) [reply scripts/build-opentracing.sh]: Use some subdir for own temp files.

Comment (jmprusi) [reply scripts/build-opentracing.sh]: going to use mktemp

File: scripts/build-opentracing.sh (hunk 4 above)
Comment (jmprusi) [reply scripts/build-opentracing.sh]: It really needs the source code.. 

edit: Not really sure, let me  try something...

File: Dockerfile.runtime [hunk 5]
```diff
@@ -4,17 +4,21 @@ FROM centos:centos7
 ARG OPENRESTY_RPM_VERSION="1.13.6.1"
 
 ENV HOME=/opt/app-root/src \
-    PATH=/opt/app-root/src/bin:/opt/app-root/bin:$PATH
+    PATH=/opt/app-root/src/bin:/opt/app-root/bin:$PATH \
+    LD_LIBRARY_PATH="$LD_LIBRARY_PATH:/opt/app-root/src/lib/"
```
Comment (mikz) [inline Dockerfile.runtime]: if `/opt/app-root` should be something like `/` of the app then the libraries should be in `/opt/app-root/lib` as `/opt/app-root/src` should have the app source.

**Comment by jmprusi** [comment]
Now travis is failing due to time constraints.. let's see if now passes.

File: scripts/build-opentracing.sh [hunk 6]
```diff
@@ -0,0 +1,113 @@
+#!/usr/bin/env bash
+
+yum install -y centos-release-scl epel-release
+
+yum install -y \
+        openresty-openssl-devel \
+        openresty-pcre-devel \
+        devtoolset-7 cmake3 pcre-devel \
+        openssl-devel thrift-devel \
```
Comment (mikz) [inline scripts/build-opentracing.sh]: Why is openssl-devel needed when openresty-openssl-devel is there too?

File: scripts/build-opentracing.sh [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+make -j"$(nproc)"
+make install
+
+#Let's build dynamic modules for opentracing and jaeger, to keep
+#binary compatibility with the openresty installed from the RPM
+#we need to configure it the same way.
+cd "${TEMP}/openresty-${OPENRESTY_RPM_VERSION}"
+
+./configure -j"$(nproc)" \
+            --with-openssl="${TEMP}/openssl/" \
```
Comment (mikz) [inline scripts/build-opentracing.sh]: openssl has patched openssl https://github.com/openresty/openresty-packaging/blob/ba9c441af3110335c3c1dbc4ae85aed6daf22c46/rpm/SPECS/openresty-openssl.spec#L13

We can't really just use the -devel package? :(

File: scripts/build-opentracing.sh (hunk 6 above)
Comment (jmprusi) [reply scripts/build-opentracing.sh]: Not really needed, removing those.

File: scripts/build-opentracing.sh (hunk 7 above)
Comment (jmprusi) [reply scripts/build-opentracing.sh]: Fixed in the last push!

**Review by mikz — APPROVED** [review]
🥇 👍 Excellent!

**Pull request merged**